hail/python/hail/hail_logging.py

```diff
@@ -0,0 +1,52 @@
+import abc
+import logging
+
+from .utils.java import Env
+
+
+class Logger(abc.ABC):
+    @abc.abstractmethod
+    def error(self, msg):
+        pass
+
+    @abc.abstractmethod
+    def warn(self, msg):
+        pass
+
+    @abc.abstractmethod
+    def info(self, msg):
+        pass
+
+
+class PythonOnlyLogger:
+    def __init__(self):
+        self.logger = logging.getLogger().getChild("hail")
+        self.logger.setLevel(logging.INFO)
```

Comment:
Why not `logging.getLogger("hail")`?

Reply:
You didn't answer this.  I'm always slightly confused about how logging configuration is supposed to work.  I think this code should be something like:

```
logging.basicConfig(level=logging.INFO)
self.logger = logging.getLogger("hail")
```

?

Reply:
I did a little digging.  `logging.getLogger().getChild("hail")` should be the same as `logging.getLogger("hail")`.

I don't think your configuration will generate output:

```
>>> import logging
>>> log = logging.getLogger('hail')
>>> log.setLevel(logging.INFO)
>>> log.info('moo')
>>>
```

Your code only turns on INFO for the hail logger. I  think I agree with that.  So I suggest this:

```
logging.basicConfig()
self.logger = logging.getLogger("hail")
self.logger.setLevel(logging.INFO)
```

Verified works:

```
>>> import logging
>>> logging.basicConfig()
>>> log = logging.getLogger('hail')
>>> log.warning('moo')
WARNING:hail:moo
>>> log.info('moo')
>>> log.setLevel(logging.INFO)
>>> log.info('moo')
INFO:hail:moo
>>> 
```

Reply:
Ah, whoops, I forgot to push the fix to use `getLogger("hail")`.

Regarding `basicConfig`, I'm pretty sure configuring that in a library is considered rude.

I do see that without basic config we don't get info statements.

```
>>> import logging
>>> x = logging.getLogger("hail")
>>> x.setLevel(logging.INFO)
>>> x.info("hi")
>>> x.warning("hi")
hi
```

Apparently when there's no handler configured, it defaults to a special handler that only prints warning and above to stderr. https://docs.python.org/3.7/howto/logging.html#what-happens-if-no-configuration-is-provided

The section on configuring logging for a library explicitly says that we shouldn't set handlers, which I believe is what basicConfig is doing. https://docs.python.org/3.7/howto/logging.html#configuring-logging-for-a-library

Ideally we would only configure if the user has not configured it, right? I guess this is hard to do because doing nothing is implicitly a configuration.

I added `basicConfig` to `context.py` but protected by the skip_logging_configuration flag.

Reply:
> The section on configuring logging for a library explicitly says that we shouldn't set handlers, which I believe is what basicConfig is doing.
 > Ideally we would only configure if the user has not configured it, right? I guess this is hard to do because doing nothing is implicitly a configuration.

`basicConfig` also includes a default format, at least:

```
$ python3
>>> import logging
>>> log = logging.getLogger('hail')
>>> log.error('foo')
foo
>>> logging.basicConfig()
>>> log.error('foo')
ERROR:hail:foo
>>> 
```

No.  Libraries should not configure.  Ideally, we wouldn't configure, and we wouldn't use logging to communicate with the user.